src/index.tsx

```diff
@@ -455,6 +456,93 @@ export class App extends React.Component<any, AppState> {
     }
   };
 
+  private getElementContainingPosition = (
+    elements: readonly ExcalidrawElement[],
+    x: number,
+    y: number,
+  ) => {
+    return getElementContainingPosition(
+      this.getElementsWithZoomScale(elements),
+      x,
+      y,
+    );
+  };
+
+  private getElementAtPosition = (
+    elements: readonly ExcalidrawElement[],
+    x: number,
+    y: number,
+  ) => {
+    return getElementAtPosition(this.getElementsWithZoomScale(elements), x, y);
+  };
+
+  private isOverScrollBars = (
+    elements: readonly ExcalidrawElement[],
+    x: number,
+    y: number,
+    canvasWidth: number,
+    canvasHeight: number,
+    scrollX: number,
+    scrollY: number,
+  ) => {
+    return isOverScrollBars(
+      this.getElementsWithZoomScale(elements),
+      x,
+      y,
+      canvasWidth,
+      canvasHeight,
+      scrollX,
+      scrollY,
+    );
+  };
+
+  private getElementsWithinSelection = (
+    elements: readonly ExcalidrawElement[],
+    selection: ExcalidrawElement,
+  ) => {
+    return getElementsWithinSelection(
+      this.getElementsWithZoomScale(elements),
+      selection,
+    );
+  };
+
+  private getCommonBounds = (elements: readonly ExcalidrawElement[]) => {
+    return getCommonBounds(this.getElementsWithZoomScale(elements));
+  };
+
+  private getElementsWithZoomScale = (
+    elements: readonly ExcalidrawElement[],
+  ): ExcalidrawElement[] => {
+    return elements.map(element => getScaledElement(element, this.state.zoom));
+  };
+
+  private setIsSelectedOnElement = (
```

Comment:
This is the strangest thing I have encountered, and I'm not sure why it was working before and why it's not anymore.

We were setting `element.isSelected = true/false` inline a bit everywhere in the code, taking advantage that `element` points to an object reference inside the `elements` array. After my changes, I found out that the selection code was still working, but `elements` would never update to reflect these inline changes. Unfortunately, the logic is way too long and not well separated, so it was impossible for me to find a reason after a couple of hours. I have literally no idea why it happens because everything I do is clone the objects inline where I need them 🤔 

So in order to move forward, I ensure the `elements` array is updated by `element.id`, which also sounds more robust to me rather than depending on object reference.

Reply:
At the beginning everything was mutable, and we moved towards making less things mutable but didn’t go all the way there.

Using the id is the solution we wanted to go. So you made the code better